docs/source/notes/gradcheck.rst

```diff
@@ -0,0 +1,227 @@
+
+Complex-to-real functions
+^^^^^^^^^^^^^^^^^^^^^^^^^
+
+To test a function :math:`g: \mathcal{C}^N \to \mathcal{R}^M, z \to y` with :math:`z = a + i b`, we reconstruct the (complex-valued) matrix that contains :math:`2 * CW`.
+
+Default complex input numerical evaluation
+""""""""""""""""""""""""""""""""""""""""""
+
+Consider the elementary case where :math:`N = M = 1` first. We know from (chapter 3 of) `this research paper <https://arxiv.org/pdf/1701.00392.pdf>`_) that:
+
+.. math::
+    CW := \frac{\partial y}{\partial z^*} = \frac{1}{2} * (\frac{\partial y}{\partial a} + i \frac{\partial y}{\partial b})
+
+Note that :math:`\frac{\partial y}{\partial a}` and :math:`\frac{\partial y}{\partial b}`, in the above equation, are :math:`\mathcal{R} \to \mathcal{R}` derivatives.
+To evaluate these numerically, we use the method described above for the real-to-real case.
+This allows us to compute the :math:`CW` matrix and then multiply it by :math:`2`.
+
+Note that the code, as of time of writing, computes this value in a slightly convoluted way:
+
+.. code:: python
+
+    # Code from https://github.com/pytorch/pytorch/blob/58eb23378f2a376565a66ac32c93a316c45b6131/torch/autograd/gradcheck.py#L99-L105
+    # Notation changes in this code block:
+    # s here is y above
+    # x, y here are a, b above
+
+    ds_dx = compute_gradient(eps)
+    ds_dy = compute_gradient(eps * 1j)
+    # conjugate wirtinger derivative
+    conj_w_d = 0.5 * (ds_dx + ds_dy * 1j)
+    # wirtinger derivative
+    w_d = 0.5 * (ds_dx - ds_dy * 1j)
+    d[d_idx] = grad_out.conjugate() * conj_w_d + grad_out * w_d.conj()
+
+    # Since grad_out is always 1, and W and CW are complex conjugate of each other, the last line ends up computing exactly `conj_w_d + w_d.conj() = conj_w_d + conj_w_d = 2 * conj_w_d`.
+
+
+Default complex input analytical evaluation
+"""""""""""""""""""""""""""""""""""""""""""
+
+Since backward mode AD computes exactly twice the :math:`CW` derivative already, we simply use the same trick as for the real-to-real case here and reconstruct the matrix row by row when there are multiple real outputs.
+
+Functions with complex outputs
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+In this case, the user-provided function does not follow the assumption from the autograd that the function we compute backward AD for is real-valued.
+This means that using autograd directly on this function is not well defined.
+To solve this, we will replace the test of the function :math:`h: \mathcal{P}^N \to \mathcal{C}^M` (where :math:`\mathcal{P}` can be either :math:`\mathcal{R}` or :math:`\mathcal{C}`), with two functions :math:`hr` and :math:`hi` such that: :math:`hr(q) = real(f(q))` and :math:`hi(q) = imag(f(q))` where :math:`q \in \mathcal{P}`.
```

Comment:
Having hi/hr definitions inline looks a bit funny. Maybe separate out the hr/hi equations into a separate block
```
To solve this, we will replace the test of the function :math:`h: \mathcal{P}^N \to \mathcal{C}^M` where math:`\mathcal{P}` can be either :math:`\mathcal{R}` or :math:`\mathcal{C}` with two functions hr and hi defined as

.. math::
    \begin{align*}
        hr(q) &:= real(f(q)) \\
        hi(q) &:= imag(f(q))
    \end{align*}

where :math:`q \in \mathcal{P}`.
..